Empty content collection when dev is first run

### Astro Info

```block
Astro                    v5.1.1
Node                     v23.3.0
System                   Windows (x64)
Package Manager          npm
Output                   static
Adapter                  none
Integrations             @astrojs/mdx
                         @astrojs/tailwind
```


### If this issue only occurs in one browser, which browser is a problem?

N/A

### Describe the Bug

On Windows, Astro reports an empty Content Layer collection when running the server with `npm run dev`. Using the built preview (`npm run build; npm run preview`) is not affected and shows all content within the Content Layer collection. I've tested the same site on Linux and it does not suffer from this issue.

### What's the expected result?

Astro should show the proper content within the Content Layer collection, instead of showing nothing.

### Link to Minimal Reproducible Example

N/A - platform-specific bug

### Participation

- [ ] I am willing to submit a pull request for this issue.

@ascorbic doing `npm i astro@experimental--dev-after-sync` didn't seem to work for me.  It installed, and `npm run dev` after didn't help.  I also tried doing the `build` and retrying `dev` to no effect.  

Here is my terminal from when i tried this with a basic blog template with no further changes after running the init script:

<details><summary>terminal</summary>

```
User@user withastro-astro-bwq4pw7d % npm i astro@experimental--dev-after-sync
npm warn ERESOLVE overriding peer dependency
npm warn While resolving: @example/blog@0.0.1
npm warn Found: astro@5.1.1
npm warn node_modules/astro
npm warn   peer astro@"^5.0.0" from @astrojs/mdx@4.0.3
npm warn   node_modules/@astrojs/mdx
npm warn     @astrojs/mdx@"^4.0.3" from the root project
npm warn   1 more (the root project)
npm warn
npm warn Could not resolve dependency:
npm warn peer astro@"^5.0.0" from @astrojs/mdx@4.0.3
npm warn node_modules/@astrojs/mdx
npm warn   @astrojs/mdx@"^4.0.3" from the root project

changed 1 package, and audited 361 packages in 1s

178 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
User@user withastro-astro-bwq4pw7d % npm run dev                             

> @example/blog@0.0.1 dev
> astro dev

08:36:07 [types] Generated 0ms
08:36:07 [content] Syncing content
08:36:07 [content] Synced content
08:36:07 [vite] Re-optimizing dependencies because lockfile has changed

 astro  v0.0.0-dev-after-sync-20241223135326 ready in 435 ms

┃ Local    http://localhost:4321/
┃ Network  use --host to expose

▶ This is a  dev-after-sync-20241223135326  prerelease build!
  Report issues here: https://astro.build/issues

08:36:07 watching for file changes...
08:36:14 [200] / 13ms
The collection "blog" does not exist or is empty. Please check your content config file for errors.
08:36:16 [200] /blog 9ms
```
</details>

--- a/.changeset/strong-scissors-sip.md
+++ b/.changeset/strong-scissors-sip.md
@@ -1,0 +1,5 @@
+---
+'astro': patch
+---
+
+Fixes a bug where content collections would sometimes appear empty when first running astro dev

--- a/packages/astro/src/content/vite-plugin-content-virtual-mod.ts
+++ b/packages/astro/src/content/vite-plugin-content-virtual-mod.ts
@@ -46,6 +46,17 @@ interface AstroContentVirtualModPluginParams {
 	fs: typeof nodeFs;
 }
 
+function invalidateDataStore(server: ViteDevServer) {
+	const module = server.moduleGraph.getModuleById(RESOLVED_DATA_STORE_VIRTUAL_ID);
+	if (module) {
+		server.moduleGraph.invalidateModule(module);
+	}
+	server.ws.send({
+		type: 'full-reload',
+		path: '*',
+	});
+}
+
 export function astroContentVirtualModPlugin({
 	settings,
 	fs,
@@ -60,7 +71,9 @@ export function astroContentVirtualModPlugin({
 		},
 		buildStart() {
 			// We defer adding the data store file to the watcher until the server is ready
-			devServer?.watcher.add(fileURLToPath(dataStoreFile));
+			devServer.watcher.add(fileURLToPath(dataStoreFile));
+			// Manually invalidate the data store to avoid a race condition in file watching
+			invalidateDataStore(devServer);
 		},
 		async resolveId(id) {
 			if (id === VIRTUAL_MODULE_ID) {
@@ -164,29 +177,17 @@ export function astroContentVirtualModPlugin({
 		configureServer(server) {
 			devServer = server;
 			const dataStorePath = fileURLToPath(dataStoreFile);
-
-			function invalidateDataStore() {
-				const module = server.moduleGraph.getModuleById(RESOLVED_DATA_STORE_VIRTUAL_ID);
-				if (module) {
-					server.moduleGraph.invalidateModule(module);
-				}
-				server.ws.send({
-					type: 'full-reload',
-					path: '*',
-				});
-			}
-
 			// If the datastore file changes, invalidate the virtual module
 
 			server.watcher.on('add', (addedPath) => {
 				if (addedPath === dataStorePath) {
-					invalidateDataStore();
+					invalidateDataStore(server);
 				}
 			});
 
 			server.watcher.on('change', (changedPath) => {
 				if (changedPath === dataStorePath) {
-					invalidateDataStore();
+					invalidateDataStore(server);
 				}
 			});
 		},

--- a/packages/astro/src/content/vite-plugin-content-virtual-mod.ts
+++ b/packages/astro/src/content/vite-plugin-content-virtual-mod.ts
@@ -70,10 +70,12 @@ export function astroContentVirtualModPlugin({
 			dataStoreFile = getDataStoreFile(settings, env.command === 'serve');
 		},
 		buildStart() {
-			// We defer adding the data store file to the watcher until the server is ready
-			devServer.watcher.add(fileURLToPath(dataStoreFile));
-			// Manually invalidate the data store to avoid a race condition in file watching
-			invalidateDataStore(devServer);
+			if (devServer) {
+				// We defer adding the data store file to the watcher until the server is ready
+				devServer.watcher.add(fileURLToPath(dataStoreFile));
+				// Manually invalidate the data store to avoid a race condition in file watching
+				invalidateDataStore(devServer);
+			}
 		},
 		async resolveId(id) {
 			if (id === VIRTUAL_MODULE_ID) {

